docs/brokerpak-specification.md

```diff
@@ -0,0 +1,334 @@
+| overwrite | boolean | If a variable already exists with the same name, should this one replace it? |
+| type | string | The JSON type of the field it will be cast to if evaluated as an expression. If defined, this MUST be a valid JSONSchema type excepting `null`. |
+
+### Example
+
+```yaml
+version: 1
+name: example-service
+id: 00000000-0000-0000-0000-000000000000
+description: a longer service description
+display_name: Example Service
+image_url: https://example.com/icon.jpg
+documentation_url: https://example.com
+support_url: https://example.com/support.html
+tags: [gcp, example, service]
+plans:
+- name: example-email-plan
+  id: 00000000-0000-0000-0000-000000000001
+  description: Builds emails for example.com.
+  display_name: example.com email builder
+  bullets:
+  - information point 1
+  - information point 2
+  - some caveat here
+  properties:
+    domain: example.com
+provision:
+  plan_inputs:
+  - required: true
+    field_name: domain
+    type: string
+    details: The domain name
+  user_inputs:
+  - required: true
+    field_name: username
+    type: string
+    details: The username to create
+  computed_inputs: []
+  template: |-
+    variable domain {type = "string"}
+    variable username {type = "string"}
+
+    output email {value = "${var.username}@${var.domain}"}
+
+  outputs:
+  - required: true
+    field_name: email
+    type: string
+    details: The combined email address
+bind:
+  plan_inputs: []
+  user_inputs: []
+  computed_inputs:
+  - name: address
+    default: ${instance.details["email"]}
+    overwrite: true
+  template: |-
+    resource "random_string" "password" {
+      length = 16
+      special = true
+      override_special = "/@\" "
+    }
+
+    output uri {value = "smtp://${var.address}:${random_string.password.result}@smtp.mycompany.com"}
+
+  outputs:
+  - required: true
+    field_name: uri
+    type: string
+    details: The uri to use to connect to this service
+examples:
+- name: Example
+  description: Examples are used for documenting your service AND as integration tests.
+  plan_id: 00000000-0000-0000-0000-000000000001
+  provision_params:
+    username: my-account
+  bind_params: {}
+
+```
+
+## Expression language reference
+
+The broker uses the [HIL expression language](https://github.com/hashicorp/hil) with a limited set of built-in functions.
+
+### Functions
+
+The following string interpolation functions are available for use:
+
+* `assert(condition_bool, message_string) -> bool`
+  * If the condition is false, then an error will be raised to the user containing `message_string`.
+  * Avoid using this function. Instead, try to make it so your users can't get into a bad state to begin with. See the "design guidelines" section.  
+  * In the words of [PEP-20](https://www.python.org/dev/peps/pep-0020/):
+    * If the implementation is hard to explain, it's a bad idea.
+    * If the implementation is easy to explain, it may be a good idea.
+* `time.nano() -> string`
```

Comment:
Could be worth calling out that this will be the server time of the node the SB is running on. I've personally have been caught by this as a developer.

Reply:
I copied in some docs from the function in Go we're wrapping. I _think_ the result should be system independent (sans some clock skew) because it's based on UTC. WDYT?

Reply:
If it's UTC, then it should be good. Might be worth mentioning that then.